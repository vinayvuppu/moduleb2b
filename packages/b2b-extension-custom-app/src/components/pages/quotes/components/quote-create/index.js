export { default } from './quote-create';
export * from './quote-create';
