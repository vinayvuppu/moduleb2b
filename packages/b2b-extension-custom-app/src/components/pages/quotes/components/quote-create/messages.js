import { defineMessages } from 'react-intl';

export default defineMessages({
  quoteSummaryTitle: {
    id: 'Quote.Create.quoteSummaryTitle',
    description: 'The title for quote summary section',
    defaultMessage: 'Quote summary',
  },
  quoteRequestedSuccess: {
    id: 'Quote.Create.quoteRequestedSuccess',
    defaultMessage: 'The quote has been created',
  },
  quoteRequestedError: {
    id: 'Quote.Create.quoteRequestedError',
    defaultMessage: 'An error has occurred',
  },
  removeVariantSuccess: {
    id: 'Quote.Create.Step.LineItems.search.removeVariantSuccess',
    defaultMessage: 'Variant has been removed successfully from the quote',
  },
  changeQuantityFailure: {
    id: 'Quote.Create.Step.LineItems.search.changeQuantityFailure',
    defaultMessage: 'An error has occurred',
  },
  changeQuantitySuccess: {
    id: 'Quote.Create.Step.LineItems.search.changeQuantitySuccess',
    defaultMessage: 'Quantity has been updated',
  },
  removeVariantFailure: {
    id: 'Quote.Create.Step.LineItems.search.removeVariantFailure',
    defaultMessage: 'An error has occurred',
  },
});
