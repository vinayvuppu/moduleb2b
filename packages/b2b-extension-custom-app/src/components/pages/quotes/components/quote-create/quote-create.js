import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import oneLineTrim from 'common-tags/lib/oneLineTrim';
import { useIntl } from 'react-intl';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { DOMAINS } from '@commercetools-frontend/constants';
import * as globalActions from '@commercetools-frontend/actions-global';
import View from '@commercetools-local/core/components/view';
import { useApplicationContext } from '@commercetools-frontend/application-shell-connectors';
import { LoadingSpinner } from '@commercetools-frontend/ui-kit';
import SaveToolbarSteps from '@commercetools-local/core/components/save-toolbar-steps';

import OrderCreateSelectStoreAndCurrency from '../../../my-company/components/order-create-select-store-and-currency';
import QuoteSummarySection from '../quote-summary-section';
import QuoteCreateAddLineItems from '../quote-create-add-line-items';

import styles from './quote-create.mod.css';
import messages from './messages';

const none = () => {};

export const QuoteCreate = props => {
  const intl = useIntl();
  const history = useHistory();

  const { currencies } = useApplicationContext(({ project }) => ({
    currencies: project.currencies,
  }));

  const [currencySelected, setCurrencySelected] = useState(
    currencies.length === 1 ? currencies[0] : undefined
  );

  const handleRemoveLineItem = async lineItemId => {
    try {
      await props.removeLineItem({ lineItemId });
      props.showNotification({
        kind: 'success',
        domain: DOMAINS.SIDE,
        text: intl.formatMessage(messages.removeVariantSuccess),
      });
    } catch (error) {
      props.showNotification({
        kind: 'error',
        domain: DOMAINS.SIDE,
        text: intl.formatMessage(messages.removeVariantFailure),
      });
    }
  };

  const handleChangeLineItem = async ({ lineItemId, quantity }) => {
    try {
      await props.changeLineItemQuantity({ lineItemId, quantity });
      props.showNotification({
        kind: 'success',
        domain: DOMAINS.SIDE,
        text: intl.formatMessage(messages.changeQuantitySuccess),
      });
    } catch (error) {
      props.showNotification({
        kind: 'error',
        domain: DOMAINS.SIDE,
        text: intl.formatMessage(messages.changeQuantityFailure),
      });
    }
  };

  const handleRequestQuote = async () => {
    try {
      props.showNotification({
        kind: 'success',
        domain: DOMAINS.SIDE,
        text: intl.formatMessage(messages.quoteRequestedSuccess),
      });
      history.push(oneLineTrim`
        /${props.projectKey}
        /b2b-extension
        /my-company
        /quotes
        /${props.quote.id}
    `);
    } catch (error) {
      props.showNotification({
        kind: 'error',
        domain: DOMAINS.SIDE,
        text: intl.formatMessage(messages.quoteRequestedError),
      });
    }
  };

  const handleCancel = async () => {
    history.push(oneLineTrim`
        /${props.projectKey}
        /b2b-extension
        /my-company
        /quotes
    `);
  };

  useEffect(() => {
    async function createQuote({
      currency,
      employeeId,
      employeeEmail,
      companyId,
    }) {
      await props.createQuote({
        currency,
        employeeId,
        companyId,
        employeeEmail,
      });
    }

    if (
      currencySelected &&
      props.company &&
      props.employee &&
      !props.quote?.id
    ) {
      createQuote({
        currency: currencySelected,
        employeeId: props.employee.id,
        employeeEmail: props.employee.email,
        companyId: props.company.id,
      });
    }

    // eslint-disable-next-line
  }, [props.company.id, props.employee.id, currencySelected]);

  if (!props.company || !props.employee) {
    return <LoadingSpinner />;
  }

  if (!currencySelected) {
    return (
      <OrderCreateSelectStoreAndCurrency
        projectKey={props.projectKey}
        goToOrdersList={none}
        updateCartDraft={({ currency }) => setCurrencySelected(currency)}
        updateStore={none}
        hideInitialSelectionModal={none}
      />
    );
  }

  return (
    <div data-track-component="Add Quote">
      <View>
        <div className={styles.container}>
          {props.quote && (
            <div className={styles.slider} data-testid="customers-list">
              <QuoteCreateAddLineItems
                quote={props.quote}
                company={props.company}
                projectKey={props.projectKey}
                addLineItem={props.addLineItem}
                renderSaveToolbarStep={() => (
                  <SaveToolbarSteps
                    currentStep={1}
                    totalSteps={1}
                    isVisible={props.quote?.lineItems.length > 0}
                    onSave={handleRequestQuote}
                    onCancel={handleCancel}
                  />
                )}
              />
            </div>
          )}
          <QuoteSummarySection
            quote={props.quote}
            onRemoveLineItem={handleRemoveLineItem}
            onChangeLineItemQuantity={handleChangeLineItem}
            title={intl.formatMessage(messages.quoteSummaryTitle)}
          />
        </div>
      </View>
    </div>
  );
};

QuoteCreate.displayName = 'QuoteCreate';
QuoteCreate.propTypes = {
  createQuote: PropTypes.func.isRequired,
  addLineItem: PropTypes.func.isRequired,
  removeLineItem: PropTypes.func.isRequired,
  changeLineItemQuantity: PropTypes.func.isRequired,
  quote: PropTypes.object,
  employee: PropTypes.object.isRequired,
  company: PropTypes.object.isRequired,
  projectKey: PropTypes.string.isRequired,
  showNotification: PropTypes.func.isRequired,
};

const mapDispatchToProps = {
  showNotification: globalActions.showNotification,
};

export default connect(null, mapDispatchToProps)(QuoteCreate);
