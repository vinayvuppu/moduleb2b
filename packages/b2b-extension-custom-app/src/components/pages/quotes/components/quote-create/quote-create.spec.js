import React from 'react';
import { shallow } from 'enzyme';
import { QuoteCreate } from './quote-create';
import QuoteSummarySection from '../quote-summary-section';
import QuoteCreateAddLineItems from '../quote-create-add-line-items';

const lineItems = [
  {
    id: 'line-item-id',
    productId: 'product-id',
    nameAllLocales: [
      {
        locale: 'en',
        value: 'Top Jucca black',
      },
    ],
    variant: {
      sku: 'M0E20000000EC8N',
      images: [
        {
          url: 'url-1',
        },
      ],
    },
    price: {
      id: 'price-id-1',
      value: {
        currencyCode: 'USD',
        centAmount: 100,
      },
    },
    quantity: 1,
    totalPrice: {
      currencyCode: 'USD',
      centAmount: 100,
    },
  },
];
const createTestProps = props => ({
  allLineItems: lineItems,
  quote: {
    id: 'quote-id',
    company: {
      id: 'company-id',
      name: 'company name',
    },
    customerGroup: {
      id: 'customer-group-id',
    },
    employeeEmail: 'employee@company.com',
    totalPrice: {
      currencyCode: 'USD',
      centAmount: 100,
    },
    lineItems,
  },
  createQuote: jest.fn(),
  addLineItem: jest.fn(),
  requestQuote: jest.fn(),
  removeLineItem: jest.fn(),
  changeLineItemQuantity: jest.fn(),
  employee: {
    id: 'employee-id',
    email: 'employee@company.com',
  },
  company: {
    id: 'company-id',
  },
  projectKey: 'project-key',
  showNotification: jest.fn(),
  ...props,
});

jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom');
  return {
    ...actual,
    useParams: jest.fn(() => ({
      projectKey: 'test-1',
    })),
    useHistory: jest.fn(() => ({ push: jest.fn() })),
  };
});

jest.mock('@commercetools-frontend/application-shell-connectors', () => {
  const actual = jest.requireActual(
    '@commercetools-frontend/application-shell-connectors'
  );
  return {
    ...actual,
    useApplicationContext: jest.fn(() => ({
      currencies: ['USD'],
    })),
  };
});

jest.mock('react-intl', () => {
  const actual = jest.requireActual('react-intl');
  return {
    ...actual,
    useIntl: jest.fn(() => ({
      formatNumber: jest.fn(({ centAmount }) => ({ centAmount })),
      formatMessage: jest.fn(() => 'formatted message'),
    })),
  };
});

describe('render', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<QuoteCreate {...props} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the QuoteCreateAddLineItems', () => {
    expect(wrapper).toRender(QuoteCreateAddLineItems);
  });
  it('should render the QuoteSummarySection', () => {
    expect(wrapper).toRender(QuoteSummarySection);
  });
});
