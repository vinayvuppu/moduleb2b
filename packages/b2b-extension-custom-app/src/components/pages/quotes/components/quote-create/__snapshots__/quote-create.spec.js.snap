// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should match snapshot 1`] = `
<div
  data-track-component="Add Quote"
>
  <View>
    <div
      className="container"
    >
      <div
        className="slider"
        data-testid="customers-list"
      >
        <withApplicationContext(injectIntl(Connect(QuoteCreateAddLineItems)))
          addLineItem={[MockFunction]}
          company={
            Object {
              "id": "company-id",
            }
          }
          projectKey="project-key"
          quote={
            Object {
              "company": Object {
                "id": "company-id",
                "name": "company name",
              },
              "customerGroup": Object {
                "id": "customer-group-id",
              },
              "employeeEmail": "employee@company.com",
              "id": "quote-id",
              "lineItems": Array [
                Object {
                  "id": "line-item-id",
                  "nameAllLocales": Array [
                    Object {
                      "locale": "en",
                      "value": "Top Jucca black",
                    },
                  ],
                  "price": Object {
                    "id": "price-id-1",
                    "value": Object {
                      "centAmount": 100,
                      "currencyCode": "USD",
                    },
                  },
                  "productId": "product-id",
                  "quantity": 1,
                  "totalPrice": Object {
                    "centAmount": 100,
                    "currencyCode": "USD",
                  },
                  "variant": Object {
                    "images": Array [
                      Object {
                        "url": "url-1",
                      },
                    ],
                    "sku": "M0E20000000EC8N",
                  },
                },
              ],
              "totalPrice": Object {
                "centAmount": 100,
                "currencyCode": "USD",
              },
            }
          }
          renderSaveToolbarStep={[Function]}
        />
      </div>
      <QuoteSummarySection
        onChangeLineItemQuantity={[Function]}
        onRemoveLineItem={[Function]}
        quote={
          Object {
            "company": Object {
              "id": "company-id",
              "name": "company name",
            },
            "customerGroup": Object {
              "id": "customer-group-id",
            },
            "employeeEmail": "employee@company.com",
            "id": "quote-id",
            "lineItems": Array [
              Object {
                "id": "line-item-id",
                "nameAllLocales": Array [
                  Object {
                    "locale": "en",
                    "value": "Top Jucca black",
                  },
                ],
                "price": Object {
                  "id": "price-id-1",
                  "value": Object {
                    "centAmount": 100,
                    "currencyCode": "USD",
                  },
                },
                "productId": "product-id",
                "quantity": 1,
                "totalPrice": Object {
                  "centAmount": 100,
                  "currencyCode": "USD",
                },
                "variant": Object {
                  "images": Array [
                    Object {
                      "url": "url-1",
                    },
                  ],
                  "sku": "M0E20000000EC8N",
                },
              },
            ],
            "totalPrice": Object {
              "centAmount": 100,
              "currencyCode": "USD",
            },
          }
        }
        title="formatted message"
      />
    </div>
  </View>
</div>
`;
